src/OrchardCore/OrchardCore.ContentManagement/FileContentDefinitionStore.cs

```diff
@@ -1,60 +1,81 @@
 
 namespace OrchardCore.ContentManagement
 {
     public class FileContentDefinitionStore : IContentDefinitionStore
     {
         private readonly IOptions<ShellOptions> _shellOptions;
-        private readonly ShellSettings _shellSettings;
+        private readonly ReaderWriterLockSlim _lock = new ReaderWriterLockSlim(LockRecursionPolicy.SupportsRecursion);
 
-        public FileContentDefinitionStore(IOptions<ShellOptions> shellOptions, ShellSettings shellSettings)
+        public FileContentDefinitionStore(IOptions<ShellOptions> shellOptions)
         {
             _shellOptions = shellOptions;
-            _shellSettings = shellSettings;
         }
 
         public Task<ContentDefinitionRecord> LoadContentDefinitionAsync()
         {
-            
-            ContentDefinitionRecord result;
+            var fileName = GetFilename();
 
-            if (!File.Exists(Filename))
+            if (!File.Exists(fileName))
             {
-                result = new ContentDefinitionRecord();
+                return Task.FromResult(new ContentDefinitionRecord());
             }
-            else
+
+            _lock.EnterReadLock();
+
+            try
             {
-                using (var file = File.OpenText(Filename))
+                using (var file = File.OpenText(fileName))
                 {
-                    var serializer = new JsonSerializer();
-                    result = (ContentDefinitionRecord)serializer.Deserialize(file, typeof(ContentDefinitionRecord));
+                    using (var reader = new JsonTextReader(file))
+                    {
+                        // We don't use 'LoadAsync' because we use a thread-affine lock type.
+                        return Task.FromResult((JObject.Load(reader)).ToObject<ContentDefinitionRecord>());
+                    }
                 }
             }
 
-            return Task.FromResult(result);
+            finally
+            {
+                _lock.ExitReadLock();
+            }
         }
 
         public Task SaveContentDefinitionAsync(ContentDefinitionRecord contentDefinitionRecord)
         {
-            using (var file = File.CreateText(Filename))
+            _lock.EnterWriteLock();
+
+            try
             {
-                var serializer = new JsonSerializer();
-                serializer.Formatting = Formatting.Indented;
-                serializer.Serialize(file, contentDefinitionRecord);
+                using (var file = File.CreateText(GetFilename()))
+                {
+                    using (var writer = new JsonTextWriter(file))
+                    {
+                        // We don't use 'WriteToAsync' because we use a thread-affine lock type.
+                        JObject.FromObject(contentDefinitionRecord).WriteTo(writer);
+                    }
+                }
+            }
+
+            finally
+            {
+                _lock.ExitWriteLock();
             }
 
             return Task.CompletedTask;
```

Comment:
I realize this wasn't changed now but still related: unless I'm misunderstanding here or this is also WIP there not much point in making such semi-async methods that actually run synchronously (it'll actually use slightly more resources). I'd also suggest not making methods async without a good reason and at least some implementations that actually run asynchronously since making everything async above it in the call stack is a huge impact.

~~In this case BTW WriteTo() could be used on ln 63.~~ No, because that'll actually throw a `NotImplementedException` :D. However, the file write can still be async (by using its `StreamWriter` directly for example. Similarly the file read in `LoadContentDefinitionAsync`.

Reply:
@Piedone thanks for your feedback. Yes i agree, in this PR the main goal is to prevent most of `GetAwaiter().GetResult()` in a given request, maybe i did too much, need to be reviewed.

I made `IContentDefinitionStore` async because the service that uses it is async and it make sense with one implementation `DatabaseContentDefinitionStore` that do an async session query.

Here for `FileContentDefinitionStore` (now used by default), i 1st did it fully async with `.LoadAsync()` and `.WriteToAsync()`. Then i realized that it was not thread safe, so i used a `ReaderWriterLockSlim` that is not async so we can't use `await` that may do a thread switching.

For a better compromise, i will see if we could use more often the new `ValueTask` type that uses less allocations, but need to learn a little more on it.

Reply:
Indeed, you're write about `ReaderWriterLockSlim`, didn't think about that. It seems [it gets really ugly once you mix it with async](https://stackoverflow.com/a/15883258/220230).

Since this file may be shared across instances in a multi-server environment (this is what happens with Azure App Services) I think a file-based locking mechanism similar to what we have in Orchard 1.x would work. While I'm uneasy by the idea actually I don't see any other simple and natural solution for this, and with `ReaderWriterLockSlim` there will still be concurrency issues.

Reply:
Yes i agree. At some point, for the same reason i also used a file based lock to update a lucene index (in fact i used indirectly the lucene one), but @sebastienros said to me to not try to make working something that is not intended to work like this.

And he said the same for `FileContentDefinitionStore`, it is not intended to be used by multiple instance sharing the same file system, in this case we need to use `DatabaseContentDefinitionStore`.

Reply:
That said, related to lucene, there is still this comment in the source code ;)

    // TODO: Lock over the filesystem in case two instances get a command to rebuild the index concurrently.

Reply:
hmm, I think there's a lot of refactoring to be made in the Lucene module. 😉

Reply:
Anyway, i now use an async `SemaphoreSlim`, even a reader can async block another reader, because now `ContentDefinitionManager` also has a tenant level cache. So, for a given tenant the definition is loaded once, until the definition is updated again.